ProjFS.Mac/PrjFSKext/VirtualizationRoots.cpp

```diff
@@ -653,3 +667,112 @@ VirtualizationRootHandle ActiveProvider_FindForPath(const char* _Nonnull path)
     
     return matchingHandle;
 }
+
+bool VirtualizationRoots_ProcessMayAccessOfflineRoots(pid_t pid)
+{
+    bool result = false;
+    
+    RWLock_AcquireShared(s_virtualizationRootsLock);
+    {
+        for (uint32_t i = 0; i < s_offlineIOPIDCount; ++i)
+        {
+            if (s_offlineIOPIDs[i] == pid)
+            {
+                result = true;
+                break;
+            }
+        }
+    }
+    RWLock_ReleaseShared(s_virtualizationRootsLock);
+
+    return result;
+}
+
+bool VirtualizationRoots_AddOfflineIOProcess(pid_t pid)
+{
+    bool success = false;
+    bool retry;
+    pid_t* resizeMemory = nullptr;
+    uint32_t resizeMemoryArrayLength = 0;
+    
```

Comment:
Same question I always ask, do we need the complexity of a resizable array here or do we want to say we can handle at most 1000 offline processes. I imagine I know the answer :-).

Reply:
If we allow registration to fail because the array is full, that failure will propagate to user space, so certain `gvfs` commands could start failing. I'm not sure that's worth a slight reduction in complexity.

Reply:
> If we allow registration to fail because the array is full, that failure will propagate to user space, so certain `gvfs` commands could start failing. I'm not sure that's worth a slight reduction in complexity.

On the flip side, if we introduce a bug in how our offline processes register/unregister it could lead to the array growing to a massive size (`UINT32_MAX / sizeof(s_offlineIOPIDs[0])`).

Users should never be running more than a handful of `gvfs commands at once` (and the functional tests are limited by the number of cores on the machine).

My preference is to pick a reasonably sized fixed-length array (e.g. 128) and that way benefit by:

- Not needing to worry about resizing the array
- Easier to detect bugs in the register/unregister flow

@jrbriggs, @jeschu1, do you have any concerns limiting ourselves in this way?

Reply:
I believe the downside of a fixed array would be that if you have over 128 offline unique pids trying to write to a unmounted repo we deny the request for those over 128.

 I don't think we should ever hit 128 and if we do my vote would be to fail, since it could be a sign of something else gone wrong.

@pmj please let me know if my understanding is incorrect

Reply:
> I don't think we should ever hit 128 and if we do my vote would be to fail, since it could be a sign of something else gone wrong.

I agree, if we find ourselves in this situation something has definitely gone wrong.

Reply:
I like the idea of just fixing the size at 128.